src/main/kotlin/app/hashers/CodeLongevity.kt

```diff
@@ -167,6 +210,57 @@ class CodeLongevity(private val serverRepo: Repo,
     }
 
     /**
+     * Scans the repo to extract code line ages.
+     */
+    fun scan() : CodeLineAges? {
+        // Load existing age data if any.
+        val istream = try { ObjectInputStream(FileInputStream(storagePath)) }
+            catch(e : Exception ) { null }
+        var storedHead = istream?.readObject() as RevCommit?
+        if (storedHead == head) {
+            return null
+        }
+
+        val ageData = (istream?.readObject() ?: CodeLineAges()) as CodeLineAges
+
+        // Update ages.
+        getLinesObservable(storedHead).blockingSubscribe { line ->
+            Logger.debug("Scanning: ${line}")
+            if (line.to.isDeleted) {
+                var age = line.age;
+                if (ageData.lastingLines.contains(line.oldId)) {
+                    age += ageData.lastingLines.remove(line.oldId)!!.age
+                }
+                val aggrAge = ageData.aggrAges.getOrPut(line.email,
+                                                        { CodeLineAges.AggrAge() } )
+                aggrAge.sum += age
+                aggrAge.count += 1
+
+            } else {
+                var age = line.age
+                if (ageData.lastingLines.contains(line.oldId)) {
+                    age += ageData.lastingLines.remove(line.oldId)!!.age
+                }
+                ageData.lastingLines.put(line.newId,
+                                         CodeLineAges.LineInfo(age, line.email))
+            }
+        }
+
+        // Store ages for subsequent runs.
+        val ostream = ObjectOutputStream(FileOutputStream(storagePath))
```

Comment:
oStream. Handle exceptions and create directories if not presented here.

Reply:
I'd rather add an empty 'storage' directory into the repo by putting .gitignore into the dir

Reply:
This directories should be created on user machine, how does it solve the problem?

Reply:
The files are currently located at 'data/storage' in the repo. As far as I understand, the problem you observed was 'storage' folder was absent. If I add one into the repo, then the problem is solved, correct?

Reply:
Not really. ML models located at src/main/resources/data/libraries/ now. What that means is that files will be packaged into JAR container. This files can't be changed during runtime. We will distribute only JAR container and all directories that used for runtime generated files should be created programmatically.

Reply:
ok, what my bet would be for a path to store the files?

Reply:
Now we only saving config file .sourcerer in user directory. I think its time to create .sourcerer/ directory instead and store all files there.

Reply:
so ./.sourcerer/data/storage or ./.sourcerer/longevity or just ./.sourcerer?

Reply:
./.sourcerer/incremental/longevity may be?

Reply:
replace incremental to more generic data?

Reply:
Ok